presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyExpressions.java

```diff
@@ -234,26 +253,63 @@ public Expression rewriteLogicalBinaryExpression(LogicalBinaryExpression node, N
                     .collect(toImmutableList());
             Expression combinedUncorrelatedPredicates = combinePredicates(node.getType(), uncorrelatedPredicates);
 
-            // Do not simplify top level conjuncts if it would result in top level disjuncts
-            // Conjuncts are easier to process when pushing down predicates.
-            if (context.isRootNode() && flippedNodeType == OR && !combinedUncorrelatedPredicates.equals(FALSE_LITERAL)) {
-                return combinePredicates(node.getType(), predicates);
-            }
-
             return combinePredicates(flippedNodeType, ImmutableList.<Expression>builder()
                     .addAll(commonPredicates)
                     .add(combinedUncorrelatedPredicates)
                     .build());
         }
 
-        private static List<List<Expression>> getSubPredicates(List<Expression> predicates)
+        private static List<List<Expression>> getSubPredicates(LogicalBinaryExpression expression)
         {
-            return predicates.stream()
+            return extractPredicates(expression.getType(), expression).stream()
                     .map(predicate -> predicate instanceof LogicalBinaryExpression ?
                             extractPredicates((LogicalBinaryExpression) predicate) : ImmutableList.of(predicate))
                     .collect(toImmutableList());
         }
 
+        /**
+         * Applies the boolean distributive property.
+         *
+         * For example:
+         * ( A & B ) | ( C & D ) => ( A | C ) & ( A | D ) & ( B | C ) & ( B | D )
+         *
+         * Returns the original expression if the expression is non-deterministic or if the distribution will
+         * create more sub-expressions than the original.
+         */
+        private static Expression distributeIfPossible(LogicalBinaryExpression expression)
```

Comment:
This is not intended to produce CNF (which needs to be applied recursively to all sub-expressions). We only replace the top level here as a simple heuristic to make sure we don't mess up our predicate move around too much with this optimization by slightly preferring conjunctions.

The only thing this is trying to do is to apply the boolean distribution property in some cases.